spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -390,7 +390,7 @@ private Expression conflictDetectionFilter() {
         filter = Expressions.and(filter, expr);
       }
 
-      return filter;
+      return SERIALIZABLE == isolationLevel ? Expressions.alwaysTrue() : filter;
```

Comment:
this is just a workaround for now so that we don't violate serializable isolation. The test fails without this

Reply:
Hm, I am not sure how it helps. Isn't the filter already true here?

@nastra, could you check if the issue is related to runtime filtering by disabling it? The source relation must be idempotent (must return the same output), which is clearly not the case here. The source may be scanned multiple times.

Reply:
Okay, I see that Spark 3.4 infers `ref(name="id") == 1005` and `ref(name="id") == 1006`. Let me think.

Reply:
Okay, the reason why we infer that predicate in Spark 3.4 is `FoldablePropagation` rule. Now let me see if we generated a wrong plan or if that rule has issues.

Reply:
The validation on the Iceberg side works correctly. The open question is whether the optimizer rule is safe to apply under these conditions and whether our Spark planning rule behaves correctly. I'll need to check with fresh eyes tomorrow.

Reply:
Here is the primary question. Are these statements equivalent?

```
MERGE INTO t USING (SELECT 1005 as id, 2 as class, SUM(value)  as value FROM t WHERE class = 1) AS u
ON t.id = u.id
WHEN MATCHED THEN UPDATE SET t.value = u.value
WHEN NOT MATCHED THEN INSERT (id, class, value) VALUES (u.id, u.class, u.value)
```

```
MERGE INTO t USING (2 as class, SUM(value)  as value FROM t WHERE class = 1) AS u
ON t.id = 1005
WHEN MATCHED THEN UPDATE SET t.value = u.value
WHEN NOT MATCHED THEN INSERT (id, class, value) VALUES (1005, u.class, u.value)
```

I think they would produce the same output.

Reply:
> The same rule applies in Spark 3.5 but the planning works differently. In Spark 3.4, we push the join condition that was already replaced. Therefore, `ref(name="id") == 1005` and `ref(name="id") == 1006` become the conflict detection conditions. According to those conditions, the operations are serializable. Therefore, both are committed.

I actually think that having `ref(name="id") == 1005` as a conflict detection condition in the one statement and `ref(name="id") == 1006` in the other statement and the way we detect conflicts isn't enough, because it can't prevent **write skew**. 

In order to avoid **write skew** (a transaction reads some data, examines the results, and takes some action based the results it saw - thus there's a causal dependency between the queries and the writes of a transaction), serializable isolation needs to detect such a case, where a TX may have acted on an outdated premise (aka the data it read may have been updated in the meantime by another TX).

What I saw in the case with Spark 3.5. is that the conflict detection filter that was pushed down to Iceberg was empty, thus we ended up using `Expressions.alwaysTrue()` and correctly detected that there's a conflict. Using `Expressions.alwaysTrue()` isn't ideal for conflict detection, but it's better than violating serializable isolation.

Reply:
In that sense, you are right that the transaction as a whole would not be serializable as it involves multiple tables. One may argue that if the source is idempotent, the operation will be considered serializable under those conflict detection conditions. 

As a data source, can we actually achieve serializability for the transactions involving reads from multiple tables without proper transaction APIs in query engines? If we modify the example to use another Iceberg table or another data source, using `true` as the conflict detection filter will not help us catch concurrent changes in other tables.

Reply:
I spent a bit more time thinking about this and I am not fully convinced using `true` as the conflict detection filter is the right approach, while I admit it works in the given example. The whole idea of the conflict detection filter is to control what is considered a serializable action against the target table (not the transaction as a whole) if there are concurrent changes to that table. In case of MERGE, we are dealing with a transaction that reads from multiple tables, so that transaction can only be serializable if we have control over both sides and can validate concurrent changes. Unfortunately, we don't have that in Spark right now as there is no transaction API.

In other words, the MERGE breaks serializability not because the conflict detection condition is wrong but because we have no way to validate concurrent changes to the other side.

Reply:
I'm 100% with you on this one @aokolnychyi, so I'm considering `true` as the conflict detection filter only as an intermediate approach. Long term we would want to fix this differently

Reply:
The proper way to ensure the MERGE statement above is serializable is to perform the following checks:
- Validate no new data matching `ref(name="id") == 1005` is added to the target table.
- Validate no new data matching `class = 1` is added to the source table.

We have the first part. I wonder if there is a workaround to validate the second part, assuming it is also an Iceberg table, without a proper transaction API in Spark. I'll need to think whether we can inject a custom rule to catch this.

Reply:
If we implement that, we can have truly serializable MERGE transactions if they touch Iceberg tables.